c_tests/run.rs

```diff
@@ -37,6 +37,7 @@ fn mk_compiler(exe: &Path, src: &Path, opt: &str) -> Command {
         "-flto",
         "-Wl,--mllvm=--embed-bitcode-final",
         "-Wl,--lto-basic-block-sections=labels",
+        "-Wl,--export-dynamic",
```

Comment:
Hang on, are we compiling *every* test as `export-dynamic`?

Reply:
[I think we should probably only compile certain tests -- presumably a test specifically designed to test this feature -- with this flag. We could probably use lang_testers `env-var` thing to set `LDFLAGS` for such a test?]

Reply:
Yes we are. I can hide it behind an env-var, but as @vext01 pointed out, we'll likely have to run everything with `export-dynamic` until we can find trace inputs at runtime.

Reply:
It's not even clear to me that using trace inputs is beneficial to this solution. It seems to me like we trade off binary size for performance when assembling the JIT module.

Reply:
Whichever solution we choose, it should be globally enabled, since any arbitrary program that we may wish to trace can access an otherwise unexported global variable.

Reply:
I must admit that I don't see what trace inputs have to do with global symbols.

Reply:
It's about finding the global variable at runtime.

If there is an exported symbol in the binary, then we don't have to do anything special, as the loader "just finds" it via standard techniques. This does mean exporting symbols that wouldn't otherwise be exposed (hence --export-dynamic). This has the advantage of being a static technique with no runtime overhead, but it will bloat the binary a little (symbols are strings).

If you don't want to export symbols, then you have to tell LLVM how to find the address of the gobal another way. One solution would be to pass a pointer to the global as a trace input. Then reading/writing from/to it is a matter of using a LLVM `load`/`store`. But this would require something like stackmaps in order to be able to find where the global. This technique avoids bloating the binary, but has a runtime cost at the time of invoking JITted trace code: you'd have to dynamically locate the global variables, probably via stackmaps (but the result could be cached).

Reply:
> It's about finding the global variable at runtime.

I understand that: I still don't understand why it relates to trace inputs.

Reply:
What @vext01 is saying is: We currently have a problem accessing global variables inside a trace. The current workaround is using `export-dynamic` so the linker can find all globals (even those that are not usually exported). In the future a better solution might be to pass the globals to the trace via trace inputs.

Reply:
> We currently have a problem accessing global variables inside a trace.

Just to check: we're not sure why we have such a problem?

Reply:
We know why we have the problem: Programs read and write global variables.

Therefore we need a way to discover the addresses of global variables.

Exporting all symbols is one such way to achieve this as it gives the binary a `name -> offset` mapping in the form of a symbol table.